src/rgw/rgw_rest_s3.cc

```diff
@@ -4410,3 +4298,192 @@ RGWOp* RGWHandler_REST_Service_S3Website::get_obj_op(bool get_data)
+    int pos = auth_str.rfind(':');
+    if (pos >=0 ) {
+      token = auth_str.substr(0, pos);
+    }
+  }
+  return token;
+}
+
+bool RGWS3V2LocalAuthEngine::is_applicable() const noexcept
+{
+  if (!RGWS3V2TokenBasedAuthEngine::is_applicable()) {
+    return false;
+  }
+
+  if (!cct->_conf->rgw_s3_auth_use_rados) {
+    return false;
+  }
+
+  return true;
+}
+
+RGWAuthApplier::aplptr_t RGWS3V2LocalAuthEngine::authenticate() const
+{
+  /* get the user info */
+  string access_key = token;
+  if (rgw_get_user_info_by_access_key(store, access_key, *(s->user)) < 0) {
+      dout(5) << "error reading user info, uid=" << token
+              << " can't authenticate" << dendl;
+      return nullptr;
+  }
+
+  /* now verify  signature */
+  string auth_hdr;
+  if (!rgw_create_s3_canonical_header(s->info, &s->header_time, auth_hdr,
+        qsr)) {
+    dout(10) << "failed to create auth header\n" << auth_hdr << dendl;
+    return nullptr;
+  }
+  dout(10) << "auth_hdr:\n" << auth_hdr << dendl;
+
+  time_t req_sec = s->header_time.sec();
+  if ((req_sec < now - RGW_AUTH_GRACE_MINS * 60 ||
+      req_sec > now + RGW_AUTH_GRACE_MINS * 60) && !qsr) {
+    dout(10) << "req_sec=" << req_sec << " now=" << now
+             << "; now - RGW_AUTH_GRACE_MINS=" << now - RGW_AUTH_GRACE_MINS * 60
+             << "; now + RGW_AUTH_GRACE_MINS=" << now + RGW_AUTH_GRACE_MINS * 60
+             << dendl;
+    dout(0)  << "NOTICE: request time skew too big now=" << utime_t(now, 0)
+             << " req_time=" << s->header_time
+             << dendl;
+    return nullptr;
+  }
+
+  map<string, RGWAccessKey>::iterator iter = s->user->access_keys.find(token);
+  if (iter == s->user->access_keys.end()) {
+    dout(0) << "ERROR: access key not encoded in user info" << dendl;
+    return nullptr;
+  }
+  RGWAccessKey& k = iter->second;
+
+  string digest;
+  int ret = rgw_get_s3_header_digest(auth_hdr, k.key, digest);
+  if (ret < 0) {
+    return nullptr;
+  }
+
+  dout(15) << "calculated digest=" << digest << dendl;
+  dout(15) << "auth_sign=" << auth_sign << dendl;
+  dout(15) << "compare=" << auth_sign.compare(digest) << dendl;
+
+  if (auth_sign != digest) {
+    return nullptr;
+  }
+
+  return apl_factory->create_apl_local(cct, *(s->user), k.subuser);
+}
+
+bool RGWS3KeystoneAuthEngine::is_applicable() const noexcept
+{
+  if (!RGWS3V2TokenBasedAuthEngine::is_applicable()) {
+    return false;
+  }
+
+  if (!cct->_conf->rgw_s3_auth_use_keystone ||
+      cct->_conf->rgw_keystone_url.empty()) {
+    return false;
+  }
+
+  return true;
+}
+
+RGWRemoteAuthApplier::acl_strategy_t RGWS3KeystoneAuthEngine::get_acl_strategy() const
+{
+  return nullptr;
+}
+
+RGWRemoteAuthApplier::AuthInfo
+RGWS3KeystoneAuthEngine::get_creds_info(const std::string& uid,
+                                  const std::string& display_name) const noexcept
+{
+  return RGWRemoteAuthApplier::AuthInfo {
+    rgw_user(uid),
+    display_name,
+    RGW_PERM_FULL_CONTROL,
+    true
```

Comment:
We're requesting admin privileges here. I'm not sure whether this the indented behavior.